File: PythonImplementation/BubbleSort.py
# This program will demonstrate a basic sorting algorithm called Bubble Sort.
# Bubble Sort is extremely easy to implement and should be used by very beginners. 
# However, Bubble Sort is not used in practice due to its runtime
# The Time Complexity for Bubble Sort is : O(n^2)
# The Space Complexity for Bubble Sort is : O(1)

def main():
    myList = [4, 6, 5, 2, 9, 3, 1, 9, 25, 13, 17, 15] # Values can be changed if desired

    sorted = False # sorted represents whether the list has been sorted or not.
    while not sorted: # While the list has not been sorted
        sorted = True # We set sort to True. If the list needs to be sorted, it will be set to False
        # Iterate through entire list. Since we are comparing with value ahead, to prevent accessing out of bounds error, we subtract our len by 1
        for i in range(len(myList) - 1):
            # If the current value is greater than the next value, a swap needs to take place. If we want to sort from greatest to least, flip the inequality
            if(myList[i] > myList[i + 1]):
                # Perform a basic swap
                temp = myList[i]
                myList[i] = myList[i+1]
                myList[i+1] = temp
                sorted = False # Since we had to enter this conditional, it means our list was not sorted completely yet.
    
    print(myList) # Print list to see result. 
# Call main
main()

File: C++Implementation/BubbleSort.cpp
#include <iostream>
using namespace std;


int main() {

	int myArray[21] = { 41, 58, 97, 53, 98, 29, 23, 19, 10, 85, 3, 90, 81, 16, 47, 78, 37, 59, 60, 43 };
	bool sorted = false; // Determines if array has been sorted, currently set to false
	// As long as array is not sorted, sort the array
	while (!sorted) {
		sorted = true; // Set sorted to true. If no change is required, will remain at true. Otherwise, it will be set to false
		// Iterate through array. We subtract by 1 to prevent index out of bounds error
		for (int i = 0; i < size(myArray) - 1; i++) {
			// If the current value is greater than next value, swap value. This is for least->greatest implementation. Flip inequality for greatest->least implementation
			if (myArray[i] > myArray[i + 1]) {
				int temp = myArray[i];
				myArray[i] = myArray[i + 1];
				myArray[i + 1] = temp;
				sorted = false;
			}
		}
	}

	for (int i = 0; i < size(myArray); i++) {
		cout << myArray[i] << endl;
	}
	return 0;
}